Fix duplicate file paths in error reporting

tosky1125 opened the pull request:
## Description

This PR fixes the issue where the same file is reported multiple times in error output when accessed through different path representations (e.g., `doc/docker.md` vs `./doc/docker.md`).

Fixes #[issue-number]

## Problem

When `lychee` discovers files through different methods (CLI arguments vs document links), the same file can be represented with different path formats:
- `doc/docker.md` (from glob pattern)
- `./doc/docker.md` (from document link)

This causes duplicate entries in the error report, making it appear as if the same error occurred in multiple files.

## Solution

Implemented path normalization in the `Display` trait for `InputSource::FsPath` to ensure consistent path representation across all platforms:

- Removes redundant `./` prefixes (including multiple occurrences like `./././`)
- Filters out intermediate `.` components (e.g., `doc/./file.md` → `doc/file.md`)
- Converts Windows backslashes to forward slashes for cross-platform consistency
- Preserves important path characteristics (absolute paths, parent directory references, trailing slashes)
- Handles edge cases like empty paths and root directories

## Changes

- Modified `Display` implementation for `InputSource` in `lychee-lib/src/types/input.rs`
- Added comprehensive test coverage for various path formats and edge cases

## Testing

Added extensive tests covering:
- [x] Basic path normalization (`./doc/file.md` → `doc/file.md`)
- [x] Multiple current directory markers (`./././doc/file.md` → `doc/file.md`)
- [x] Windows path normalization (`doc\file.md` → `doc/file.md`)
- [x] Cross-platform consistency
- [x] Edge cases (empty paths, trailing slashes, absolute paths)
- [x] All existing tests pass

## Notes

- The normalization only affects display output, not the actual file processing logic
- Maintains backward compatibility with existing behavior
- No performance impact as normalization only occurs during error reporting

mre reviewed (commented):
Interesting idea. 

- The issue name is not filled out in the pull request description. 
- There are a few comments in the PR code, which I would clean up. Will highlight them. 
- Needs an Integration test in cli.rs
- Does Windows path conversion in multiple places. I'm also not sure if there aren't any edge cases with that. Maybe there's a library for it. Also wondering if we even need to change the Windows paths because I think it's good to know the exact path that was checked, even if that means the output is different on different platforms. 
- I think the path cleanup should be in a separate method (or even module). Display should not have a lot of business logic, but just call into other methods if needed.

tosky1125 commented:
@mre Thank you for the thorough review! I've addressed most of your feedback:

### Changes made:
1.  **Extracted normalization logic**: Moved path normalization from `Display` impl to a dedicated `InputSource::normalize_path_display()` method for better separation of concerns.

2. **Added integration test**: Added `test_normalized_path_display` in `cli.rs` to verify that paths like `./test.md` are displayed consistently as `test.md`.

3. **Removed Windows path conversion**: As you suggested, I've removed the backslash-to-forward-slash conversion to preserve platform-specific paths. Now only `.` normalization is performed.

4. **Cleaned up comments**: Removed redundant comments and moved variable declarations closer to their usage.

5. **Simplified implementation**: The current implementation only removes leading `./` for consistency, avoiding complex path manipulations.

Additionally, I've confirmed that the CI failures are unrelated to this PR:

The failing tests expect:
- `successful: 10`
- `errors: 0`

But actually get:
- `successful: 7` 
- `errors: 1` (403 from creativecommons.org)

Our path normalization only affects relative paths (removes `./` prefix) and preserves absolute paths unchanged. The JSON serialization works correctly with our changes.

This is purely an external service issue affecting all PRs.

mre commented on lychee-lib/src/types/input.rs:
```diff
@@ -88,15 +88,50 @@ impl Serialize for InputSource {
+        if path_str == "./" {
+            "./".to_string()
+        } else if has_trailing_slash && result != "." && !result.ends_with('/') {
+            format!("{result}/") // 수정된 부분
```
Comment can be removed

tosky1125 replied:
removed 👍

thomas-zahner commented:
Thank you for the PR. This is an interesting idea. However, this does not really fix https://github.com/lycheeverse/lychee/issues/1660 as to my understanding. This now "normalises" output files to some specific format. But the affected files are still checked multiple times and also displayed multiple times. To really address the issue we would need to filter out such duplicates before checking them.

thomas-zahner commented on lychee-bin/tests/cli.rs:
```diff
@@ -2222,4 +2222,24 @@ mod cli {
+
+        // Both should be displayed as "test.md" (not "./test.md")
+        assert!(stdout.contains("[test.md]:"));
+        assert!(!stdout.contains("[./test.md]:"));
```
To really fix https://github.com/lycheeverse/lychee/issues/1660 as to my understanding the assertions must check that `test.md` (or `./test.md`) occurs exactly once in `stdout`. In the current state of the PR this assertion will fail because `test.md` occurs twice.

mre commented:
> this does not really fix https://github.com/lycheeverse/lychee/issues/1660 

That is correct. @tosky1125, can you remove the `Fixes #1660` from the description? I think the PR is still valuable to be merged as is because it cleans up the output a bit. @thomas-zahner, do you agree?

thomas-zahner commented on lychee-bin/600c0d6d-eacd-4fba-bef0-bd79a4da72c2.json:
```diff
@@ -0,0 +1,27 @@
+{
```
What's the purpose of all these JSON files? They don't seem to be used in the test, right?

thomas-zahner replied:
Realised that these JSON files were likely created with the `test_json_output` macro after test failures. This was because the files were not cleaned up after a failing assertion. I've fixed that behaviour with https://github.com/lycheeverse/lychee/pull/1771/commits/ec4bedb3d4d21ecc0a2db9b6f94423d4f43ca8ba

thomas-zahner commented on lychee-lib/src/types/input.rs:
```diff
@@ -88,15 +88,50 @@ impl Serialize for InputSource {
-            Self::Stdin => "stdin",
-            Self::String(s) => s,
-        })
+        match self {
```
I'd much prefer the previous position of `match` as this didn't duplicate `f.write_str` five times

thomas-zahner reviewed (commented):
After giving it some thought, I came to the conclusion that I would prefer not to merge this. As mentioned earlier addressing the root cause to eliminate duplicates would be great, but this PR does not solve that. 

This PR applies some cosmetic changes to the output. However, I actually prefer the current output because it contains more information.

Given the following example `cargo run *.md ./*.md --dump-inputs` we currently get

```
CONTRIBUTING.md
README.md
./CONTRIBUTING.md
./README.md
```

with this PR I would expect the following (noticed that it's not yet the case because this PR only affects normal link checking as of now, not dumping [which would be another reason for not merging the PR because I expect consistent behaviour as a user across all commands])

```
CONTRIBUTING.md
README.md
CONTRIBUTING.md
README.md
```

This removes some useful information in my opinion. I no longer understand as easily why my files were included twice for link checking.

mre commented:
I went ahead and closed this as it doesn't resolve an immediate problem. Thanks anyway for the idea and the PR, @tosky1125.

thomas-zahner commented:
Yes, thank you nonetheless for the idea and your efforts @tosky1125.
